pallets/farming/src/gauge.rs

```diff
@@ -0,0 +1,494 @@
+                            reward_to_claim,
+                        )
+                    },
+                )?;
+                gauge_info.last_claim_block = current_block_number;
+                gauge_info.claimed_time_factor = latest_claimed_time_factor;
+                if gauge_info.gauge_stop_block <= current_block_number {
+                    let ed = T::MultiCurrency::minimum_balance(gauge_pool_info.token);
+                    let mut account_to_send = who.clone();
+
+                    if gauge_info.gauge_amount < ed {
+                        let receiver_balance =
+                            T::MultiCurrency::total_balance(gauge_pool_info.token, who);
+
+                        let receiver_balance_after = receiver_balance
+                            .checked_add(&gauge_info.gauge_amount)
+                            .ok_or(ArithmeticError::Overflow)?;
+                        if receiver_balance_after < ed {
+                            account_to_send = T::TreasuryAccount::get();
+                        }
+                    }
+                    T::MultiCurrency::transfer(
+                        gauge_pool_info.token,
+                        &gauge_pool_info.keeper,
+                        &account_to_send,
+                        gauge_info.gauge_amount,
+                    )?;
+                    gauge_pool_info.total_time_factor = gauge_pool_info
+                        .total_time_factor
+                        .checked_sub(gauge_info.total_time_factor)
+                        .ok_or(ArithmeticError::Overflow)?;
+                    gauge_pool_info.gauge_amount = gauge_pool_info
+                        .gauge_amount
+                        .checked_sub(&gauge_info.gauge_amount)
+                        .ok_or(ArithmeticError::Overflow)?;
+                } else {
+                    *maybe_gauge_info = Some(gauge_info);
+                };
+                GaugePoolInfos::<T>::insert(gid, gauge_pool_info);
+            }
+            Ok(())
+        })?;
+        Ok(())
+    }
+
+    pub fn get_farming_rewards(
+        who: &T::AccountId,
+        pid: PoolId,
+    ) -> Result<RewardOf<T>, DispatchError> {
+        let share_info =
+            SharesAndWithdrawnRewards::<T>::get(pid, who).ok_or(Error::<T>::ShareInfoNotExists)?;
+        let pool_info = PoolInfos::<T>::get(pid).ok_or(Error::<T>::PoolDoesNotExist)?;
+        let total_shares = U256::from(pool_info.total_shares.to_owned().saturated_into::<u128>());
+        let mut result_vec = Vec::<(CurrencyIdOf<T>, BalanceOf<T>)>::new();
+
+        pool_info.rewards.iter().try_for_each(
+            |(reward_currency, (total_reward, total_withdrawn_reward))| -> DispatchResult {
+                let withdrawn_reward = share_info
+                    .withdrawn_rewards
+                    .get(reward_currency)
+                    .copied()
+                    .unwrap_or_default();
+
+                let total_reward_proportion: BalanceOf<T> =
```

Comment:
I feel confused about the way to calculate portion, Why is it a integer instead if percent?
Dividing `total_shares` should return a a float number (0, 1), right?

Reply:
PS: also refactor the reward algorithms into lib.rs